### PR TITLE
feat: filter the stacks by app; display git info

### PR DESCRIPTION
**Pull request by jakeyheath**
This PR changes the behavior of `happy list` so that it only shows the stacks that match the current app you are currently using. There is an optional `--all` flag that can be used to show all stacks in the same Happy environment. Additionally, there are some changes to the fields that are displayed in the list table. These changes include the app name, Git branch, the Git SHA (if the user deployed a clean git tree), the Github repo and abbreviated versions of the timestamp.

In order for people to use this, we need to have two version of the StackMeta struct. One for the new features and one for the legacy format. I have marked all the legacy code as DEPRECATED and we should clean up going forward. Here is an example of what the output will look like:

Screenshot 2023-04-19 at 12 19 31 PM

For legacy stacks, it will look something like this:

Screenshot 2023-04-19 at 12 21 00 PM

As stacks get updated with the latest version of Happy CLI, these fields will fill in.

### CODE REVIEW

File: cli/pkg/output/printer_factory.go [hunk 1]
```diff
@@ -37,7 +38,10 @@ func NewPrinter(outputFormat string) Printer {
 type StackConsoleInfo struct {
 	Name        string `header:"Name"`
 	Owner       string `header:"Owner"`
-	Tag         string `header:"Tags"`
+	App         string `header:"App"`
+	Repo        string `header:"Repo"`
+	GitBranch   string `header:"GitBranch"`
+	GitSHA      string `header:"GitSHA"`
```
Comment (alexlokshin-czi) [inline cli/pkg/output/printer_factory.go:44]:
```suggestion
GitSHA      string `header:"SHA"`
```

File: cli/pkg/output/printer_factory.go [hunk 2]
```diff
@@ -37,7 +38,10 @@ func NewPrinter(outputFormat string) Printer {
 type StackConsoleInfo struct {
 	Name        string `header:"Name"`
 	Owner       string `header:"Owner"`
-	Tag         string `header:"Tags"`
+	App         string `header:"App"`
+	Repo        string `header:"Repo"`
+	GitBranch   string `header:"GitBranch"`
```
Comment (alexlokshin-czi) [inline cli/pkg/output/printer_factory.go:43]:
```suggestion
	GitBranch   string `header:"Branch"`
```

File: cli/pkg/output/printer_factory.go [hunk 3]
```diff
@@ -59,16 +63,30 @@ func Stack2Console(ctx context.Context, stack stackservice.StackInfo) StackConso
 		uniqueMap[endpoint] = true
 	}
 	for endpoint := range uniqueMap {
+		// filter out the k8s cluster endpoints
+		if strings.Contains(endpoint, "svc.cluster") {
```
Comment (alexlokshin-czi) [inline cli/pkg/output/printer_factory.go:67]: That's cheating ;)

File: cli/pkg/output/printer_factory.go [hunk 4]
```diff
@@ -59,16 +63,30 @@ func Stack2Console(ctx context.Context, stack stackservice.StackInfo) StackConso
...
+			continue
+		}
 		endpoints = append(endpoints, endpoint)
 	}
 
+	abbrevRepo := strings.TrimPrefix(strings.TrimSuffix(stack.Repo, ".git"), "git@github.com:")
+	abbrevOwner := strings.TrimSuffix(stack.Owner, "@chanzuckerberg.com")
```
Comment (alexlokshin-czi) [inline cli/pkg/output/printer_factory.go:74]: We need to make this configurable in the future

File: cli/pkg/stack_mgr/stack_service.go [hunk 5]
```diff
@@ -356,6 +319,57 @@ func (s *StackService) GetStacks(ctx context.Context) (map[string]*Stack, error)
...
+	stacks, err := s.GetStacks(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Iterate in order
+	stackNames := maps.Keys(stacks)
+	stackInfos := make([]*StackInfo, len(stackNames))
+	sort.Strings(stackNames)
+	for i, name := range stackNames {
+		i, name := i, name // https://golang.org/doc/faq#closures_and_goroutines
```
Comment (alexlokshin-czi) [inline cli/pkg/stack_mgr/stack_service.go:333]: Nice!

Comment (jakeyheath) [reply cli/pkg/stack_mgr/stack_service.go:333]: ya I noticed when there was a lot of stacks, it was taking a uncomfortably long time to make the list. This should help. I think we have a lot of improvements to make regarding concurrency.

File: cli/pkg/output/printer_factory.go (hunk 4 above)
Comment (jakeyheath) [reply cli/pkg/output/printer_factory.go:74]: yea

File: shared/workspace_repo/workspace.go [hunk 6]
```diff
@@ -616,14 +641,14 @@ func (s *TFEWorkspace) GetEndpoints(ctx context.Context) (map[string]string, err
 	endpoints := map[string]string{}
 	outputs, err := s.GetOutputs(ctx)
 	if err != nil {
-		return nil, errors.Wrap(err, "Unable to get workspace outputs")
+		return nil, errors.Wrap(err, "unable to get workspace outputs")
 	}
 	if endpoint, ok := outputs["frontend_url"]; ok {
 		endpoints["FRONTEND"] = endpoint
-	} else if svc_endpoints, ok := outputs["service_endpoints"]; ok {
+	} else if svc_endpoints, ok := outputs["service_urls"]; ok {
```
Comment (alexlokshin-czi) [inline shared/workspace_repo/workspace.go:648]: Is this going to be a problem with all the existing repos?

File: cli/pkg/stack_mgr/stack_meta.go [hunk 7]
```diff
@@ -1,126 +1,246 @@
...
+		isClean, _, err := util.IsCleanGitTree(dir)
+		if err != nil {
+			logrus.Errorf("error checking if git tree in %s was clean: %s", dir, err)
+			return
+		}
+		if !isClean {
+			s.GitSHA = "dirty git tree (PLEASE COMMIT YOUR CHANGES)"
+			return
+		}
+		cmd := exec.Command("git", "rev-parse", "--short", "HEAD")
```
Comment (alexlokshin-czi) [inline cli/pkg/stack_mgr/stack_meta.go:202]: We need to check git executable is in the path...

File: shared/workspace_repo/workspace.go (hunk 6 above)
Comment (jakeyheath) [reply shared/workspace_repo/workspace.go:648]: No, the only place where service_endspoints was used was our examples lol

Comment (jakeyheath) [reply shared/workspace_repo/workspace.go:648]: Everywhere else has service_urls

**Review by alexlokshin-czi — APPROVED** [review]

**Pull request merged**